src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs

```diff
@@ -90,7 +90,7 @@ public void Extract_LinkEntry_TargetOutsideDirectory(TarEntryType entryType)
 
             using TempDirectory root = new TempDirectory();
 
-            Assert.Throws<IOException>(() => TarFile.ExtractToDirectory(archive, root.Path, overwriteFiles: false));
+            Assert.ThrowsAny<IOException>(() => TarFile.ExtractToDirectory(archive, root.Path, overwriteFiles: false));
```

Comment:
Why is this `ThrowsAny` now?

Reply:
I assume the exception gets thrown from a different code path. I'll re-run the test and see how it changed.

Reply:
Previously, we detected this case and manually threw an `IOException`. With the changes, a `FileNotFoundException` is thrown when the link is created:

```
[xUnit.net 00:00:02.05]     System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.Extract_LinkEntry_TargetOutsideDirectory(entryType: HardLink) [FAIL]
  Failed System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.Extract_LinkEntry_TargetOutsideDirectory(entryType: HardLink) [140 ms]
  Error Message:
   Assert.Throws() Failure
Expected: typeof(System.IO.IOException)
Actual:   typeof(System.IO.FileNotFoundException): Could not find file '/tmp/unzcxv13.n4w/link'.
---- System.IO.FileNotFoundException : Could not find file '/tmp/unzcxv13.n4w/link'.
  Stack Trace:
     at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError) in /home/tmds/repos/runtime/src/libraries/Common/src/Interop/Unix/Interop.IOErrors.cs:line 16
   at Interop.CheckIo(Int64 result, String path, Boolean isDirError) in /home/tmds/repos/runtime/src/libraries/Common/src/Interop/Unix/Interop.IOErrors.cs:line 42
   at Interop.CheckIo(Int32 result, String path, Boolean isDirError) in /home/tmds/repos/runtime/src/libraries/Common/src/Interop/Unix/Interop.IOErrors.cs:line 66
   at System.Formats.Tar.TarEntry.ExtractAsHardLink(String targetFilePath, String hardLinkFilePath) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.Unix.cs:line 40
   at System.Formats.Tar.TarEntry.CreateNonRegularFile(String filePath, String linkTargetPath) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:line 454
   at System.Formats.Tar.TarEntry.ExtractToFileInternal(String filePath, String linkTargetPath, Boolean overwrite) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:line 399
   at System.Formats.Tar.TarEntry.ExtractRelativeToDirectory(String destinationDirectoryPath, Boolean overwrite, SortedDictionary`2 pendingModes) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:line 300
   at System.Formats.Tar.TarFile.ExtractToDirectoryInternal(Stream source, String destinationDirectoryPath, Boolean overwriteFiles, Boolean leaveOpen) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:line 458
   at System.Formats.Tar.TarFile.ExtractToDirectory(Stream source, String destinationDirectoryName, Boolean overwriteFiles) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:line 190
   at System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.<>c__DisplayClass5_0.<Extract_LinkEntry_TargetOutsideDirectory>b__0() in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs:line 93
----- Inner Stack Trace -----
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError) in /home/tmds/repos/runtime/src/libraries/Common/src/Interop/Unix/Interop.IOErrors.cs:line 16
   at Interop.CheckIo(Int64 result, String path, Boolean isDirError) in /home/tmds/repos/runtime/src/libraries/Common/src/Interop/Unix/Interop.IOErrors.cs:line 42
   at Interop.CheckIo(Int32 result, String path, Boolean isDirError) in /home/tmds/repos/runtime/src/libraries/Common/src/Interop/Unix/Interop.IOErrors.cs:line 66
   at System.Formats.Tar.TarEntry.ExtractAsHardLink(String targetFilePath, String hardLinkFilePath) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.Unix.cs:line 40
   at System.Formats.Tar.TarEntry.CreateNonRegularFile(String filePath, String linkTargetPath) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:line 454
   at System.Formats.Tar.TarEntry.ExtractToFileInternal(String filePath, String linkTargetPath, Boolean overwrite) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:line 399
   at System.Formats.Tar.TarEntry.ExtractRelativeToDirectory(String destinationDirectoryPath, Boolean overwrite, SortedDictionary`2 pendingModes) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:line 300
   at System.Formats.Tar.TarFile.ExtractToDirectoryInternal(Stream source, String destinationDirectoryPath, Boolean overwriteFiles, Boolean leaveOpen) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:line 458
   at System.Formats.Tar.TarFile.ExtractToDirectory(Stream source, String destinationDirectoryName, Boolean overwriteFiles) in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:line 190
   at System.Formats.Tar.Tests.TarFile_ExtractToDirectory_Stream_Tests.<>c__DisplayClass5_0.<Extract_LinkEntry_TargetOutsideDirectory>b__0() in /home/tmds/repos/runtime/src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs:line 93

```

Reply:
In that case we should probably update to:
```suggestion
            Assert.Throws<FileNotFoundException>(() => TarFile.ExtractToDirectory(archive, root.Path, overwriteFiles: false));
```

Reply:
Huh actually this keeps failing with IOException on the Symlink case.